### PR TITLE
feat: Support breadcrumb instances deduplication

### PR DESCRIPTION
**Pull request by just-boris**
### Description

Deduplicate multiple `BreadcrumbGroup` instances across the page and render only one inside the respective app layout slot. Only behind `appLayoutWidget` feature flag, not for production use.

```js
<AppLayout
   breadcrumbs={null}
   content={<>
       <BreadcrumbGroup /> // this one will be rendered in the intended place, even if it did not use the slot
   </>}
/>
```

See the tests for more details about order resolution and other corner-cases.

Related links, issue #, if available: n/a

### How has this been tested?

Added unit and integration tests

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/app-layout/visual-refresh-toolbar/index.tsx [hunk 1]
```diff
@@ -176,9 +177,15 @@ const AppLayoutVisualRefreshToolbar = React.forwardRef(
       splitPanelPosition: splitPanelPreferences?.position,
     });
 
+    const discoveredBreadcrumbs = useGetGlobalBreadcrumbs();
+
     const hasToolbar =
       !embeddedViewMode &&
-      (!!resolvedNavigation || !!breadcrumbs || splitPanelToggleConfig.displayed || drawers!.length > 0);
+      (!!resolvedNavigation ||
+        !!breadcrumbs ||
```
Comment (just-boris) [inline src/app-layout/visual-refresh-toolbar/index.tsx:185]: F

File: src/app-layout/visual-refresh-toolbar/toolbar/index.tsx [hunk 2]
```diff
@@ -106,8 +108,11 @@ export function AppLayoutToolbarImplementation({ appLayoutInternals }: AppLayout
             />
           </nav>
         )}
-        {breadcrumbs && (
-          <div className={clsx(styles['universal-toolbar-breadcrumbs'], testutilStyles.breadcrumbs)}>{breadcrumbs}</div>
+        {(breadcrumbs || discoveredBreadcrumbs) && (
+          <div className={clsx(styles['universal-toolbar-breadcrumbs'], testutilStyles.breadcrumbs)}>
+            {breadcrumbs}
+            {discoveredBreadcrumbs && <InternalBreadcrumbGroup {...discoveredBreadcrumbs} />}
+          </div>
```
Comment (just-boris) [inline src/app-layout/visual-refresh-toolbar/toolbar/index.tsx:115]: The `breadcrumbs` value here is also deduplicated into `discoveredBreadcrumbs` same way as any other instance on the page. No special treatment for "own breadcrumbs"

File: src/breadcrumb-group/index.tsx [hunk 3]
```diff
@@ -5,14 +5,21 @@ import { BreadcrumbGroupProps } from './interfaces';
...
+import { useSetGlobalBreadcrumbs } from '../internal/plugins/helpers/use-global-breadcrumbs';
 
 export { BreadcrumbGroupProps };
 
 export default function BreadcrumbGroup<T extends BreadcrumbGroupProps.Item = BreadcrumbGroupProps.Item>({
   items = [],
   ...props
 }: BreadcrumbGroupProps<T>) {
+  const registeredGlobally = useSetGlobalBreadcrumbs({ items, ...props });
   const baseComponentProps = useBaseComponent('BreadcrumbGroup');
```
Comment (just-boris) [inline src/breadcrumb-group/index.tsx:17]: Usage metrics are still emitted for each instance individually

File: src/internal/plugins/helpers/use-global-breadcrumbs.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  });
+
+  return registered;
+}
+
+export function useSetGlobalBreadcrumbs<T extends BreadcrumbGroupProps.Item>(props: BreadcrumbGroupProps<T>) {
+  // avoid additional side effects when this feature is not active
+  if (!getGlobalFlag('appLayoutWidget')) {
+    return false;
+  }
```
Comment (just-boris) [inline src/internal/plugins/helpers/use-global-breadcrumbs.ts:35]: This guard prevents any changes for the current production behavior

File: src/internal/plugins/helpers/use-global-breadcrumbs.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    registrationRef.current = registration;
+
+    return () => {
+      registration.cleanup();
+    };
+    // subsequent prop changes are handled by another effect
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  useLayoutEffect(() => {
```
Comment (connorlanigan) [inline src/internal/plugins/helpers/use-global-breadcrumbs.ts:24]: Since `useLayoutEffect` can have a negative performance impact, (why) do we need to use it?

File: src/app-layout/visual-refresh-toolbar/index.tsx [hunk 6]
```diff
@@ -176,9 +177,15 @@ const AppLayoutVisualRefreshToolbar = React.forwardRef(
...
+        !!discoveredBreadcrumbs ||
```
Comment (connorlanigan) [inline src/app-layout/visual-refresh-toolbar/index.tsx:186]: I think we don't need the double exclamation marks here, and they make the code difficult to read. We could make this block more readable by removing them, and converting only the result to a boolean:
```
const hasToolbar = Boolean(
      !embeddedViewMode &&
      (resolvedNavigation ||
        breadcrumbs ||
        discoveredBreadcrumbs ||
        splitPanelToggleConfig.displayed ||
        drawers!.length > 0)
);
```

File: pages/app-layout/global-breadcrumbs.page.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+// SPDX-License-Identifier: Apache-2.0
+import React, { useState } from 'react';
+import AppLayout from '~components/app-layout';
+import SpaceBetween from '~components/space-between';
+import labels from './utils/labels';
+import BreadcrumbGroup from '~components/breadcrumb-group';
+
+export default function () {
+  const [extraBreadcrumb, setExtraBreadcrumb] = useState(false);
+  return (
```
Comment (connorlanigan) [inline pages/app-layout/global-breadcrumbs.page.tsx:11]: Maybe we could add a `<p>` to inform that this page only works when the `appLayoutWidget` flag is enabled (and maybe its current status?) - it could be confusing otherwise, making people think the feature is broken

File: src/internal/plugins/controllers/breadcrumbs.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    this.#appLayoutUpdateCallback(latestBreadcrumb?.props ?? null);
+  }, 0);
+
+  #notifyBreadcrumbs = debounce(() => {
+    if (!this.#appLayoutUpdateCallback) {
+      return;
+    }
+    this.#breadcrumbRegistrations.forEach(listener => listener());
+  }, 0);
+
```
Comment (connorlanigan) [inline src/internal/plugins/controllers/breadcrumbs.ts:41]: We handle multiple breadcrumb instances, do we also need to handle multiple AppLayouts?

Comment (just-boris) [reply src/internal/plugins/controllers/breadcrumbs.ts:41]: Handled below. Only one app layout can register as breadcrumbs holder

File: src/internal/plugins/helpers/use-global-breadcrumbs.ts (hunk 5 above)
Comment (just-boris) [reply src/internal/plugins/helpers/use-global-breadcrumbs.ts:24]: It causes impact when touches the hosting component:


```
useLayoutEffect(() => {
   setState(something); // this is bad
}, [...])
```

Here we call an API outside React, no concerns here.

File: src/app-layout/visual-refresh-toolbar/index.tsx (hunk 6 above)
Comment (just-boris) [reply src/app-layout/visual-refresh-toolbar/index.tsx]: Applied

**Review by connorlanigan — APPROVED** [review]
